### DISCUSSION TITLE
Connection string in oracle 19.x and 18.x

### DISCUSSION THREAD

**Discussion by tandem97** [godror/godror#181] [Q&A] [answered]
```go
var connectString = fmt.Sprintf(`user="%v" password="%v" connectString="%v:%v/%v?connect_timeout=5" noTimezoneCheck=true`, dbuser, dbpassword, dbhost, dbport, dbservicename)
	newDbdataSource.db, err = sql.Open("godror", connectString)
```
This is my connection to oracle that i was using  with oracle client version 19.12.0.0.0 basic package and everything were good. But now i need my go utility run on win 7 host. Oracle client  version 19.12.0.0.0 does not support win7, so i downloaded previos version 18.5.0.0.0. With this version i cant work with db and have errors like "service with such name is not fount". When i delete "?connect_timeout=5" from connectString everything is working but i dont want to refuse timeout. I cant understand how to set connect_timeout with this oracle client version. If some one have the same problem or know how to solve it i will be appreciated

**Comment by sudarshan12s** [comment]
> ```go
> var connectString = fmt.Sprintf(`user="%v" password="%v" connectString="%v:%v/%v?connect_timeout=5" noTimezoneCheck=true`, dbuser, dbpassword, dbhost, dbport, dbservicename)
> 	newDbdataSource.db, err = sql.Open("godror", connectString)
> ```
> 
> This is my connection to oracle that i was using with oracle client version 19.12.0.0.0 basic package and everything were good. But now i need my go utility run on win 7 host. Oracle client version 19.12.0.0.0 does not support win7, so i downloaded previos version 18.5.0.0.0. With this version i cant work with db and have errors like "service with such name is not fount". When i delete "?connect_timeout=5" from connectString everything is working but i dont want to refuse timeout. I cant understand how to set connect_timeout with this oracle client version. If some one have the same problem or know how to solve it i will be appreciated

I could not verify on 18.5. Does sqlnet.ora settings work for you? sqlnet.ora is present in $TNS_ADMIN folder path. you can add below entry like this. unit is in secs by default. 
```SQLNET.OUTBOUND_CONNECT_TIMEOUT=5 ```

you can also mention path of sqlnet.ora using configDir=c:\Users\tnsadminpath\ in connectstring. 

```connectString="ip:1521/svcname" configDir=c:\User\tnsadminpath\```

**Reply by tandem97** [reply]
> > ```go
> > var connectString = fmt.Sprintf(`user="%v" password="%v" connectString="%v:%v/%v?connect_timeout=5" noTimezoneCheck=true`, dbuser, dbpassword, dbhost, dbport, dbservicename)
> > 	newDbdataSource.db, err = sql.Open("godror", connectString)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is my connection to oracle that i was using with oracle client version 19.12.0.0.0 basic package and everything were good. But now i need my go utility run on win 7 host. Oracle client version 19.12.0.0.0 does not support win7, so i downloaded previos version 18.5.0.0.0. With this version i cant work with db and have errors like "service with such name is not fount". When i delete "?connect_timeout=5" from connectString everything is working but i dont want to refuse timeout. I cant understand how to set connect_timeout with this oracle client version. If some one have the same problem or know how to solve it i will be appreciated
> 
> I could not verify on 18.5. Does sqlnet.ora settings work for you? sqlnet.ora is present in $TNS_ADMIN folder path. you can add below entry like this. unit is in secs by default.
> `SQLNET.OUTBOUND_CONNECT_TIMEOUT=5 `
> 
> you can also mention path of sqlnet.ora using configDir=c:\Users\tnsadminpath\ in connectstring.
> 
> `connectString="ip:1521/svcname" configDir=c:\User\tnsadminpath\`

I just  download instant client from https://www.oracle.com/database/technologies/instant-client/winx64-64-downloads.html and put files with my utility. I cant install client on the host, so i dont have tnsnames and etc.

**Reply by sudarshan12s** [reply]
new sqlnet.ora file alone can be created with above entry in some folder accessible to godror client.  The path can be mentioned in ConfigDir.

**Reply by tandem97** [reply]
> new sqlnet.ora file alone can be created with above entry in some folder accessible to godror client. The path can be mentioned in ConfigDir.

I need to create this file and just add 1 string ```SQLNET.OUTBOUND_CONNECT_TIMEOUT=5``` ?

**Reply by sudarshan12s** [reply]
yes...

**Comment by cjbj** [comment]
"Easy Connect Plus" syntax like "...?connect_timeout=5" is a [19c feature](https://download.oracle.com/ocomdocs/global/Oracle-Net-19c-Easy-Connect-Plus.pdf), so won't work with 18c client libraries.  You need to use sqlnet.ora in that version as mentioned.

**Comment by tandem97** [comment, marked as answer]
I read some more docs and found that i can use tnsname's connection string syntax.
```go
connectString := fmt.Sprintf(`user="%v" password="%v" connectString="(DESCRIPTION=(connect_timeout=5)(ADDRESS=(PROTOCOL=TCP)(HOST=%v)(PORT=%v))(CONNECT_DATA=(SERVICE_NAME=%v)))" noTimezoneCheck=true`, dbuser, dbpassword, dbhost, dbport, dbservicename)
```
It works with both oracle clients versions

**Reply by sudarshan12s** [reply]
I checked providing with above format (DESCRIPTION= ) have parsing for connect_timeout . It is same as when you alternatively provide this format in tnsnames.ora. parsing is not there with Easy connect plus in 18.5 but its there from 19.2
